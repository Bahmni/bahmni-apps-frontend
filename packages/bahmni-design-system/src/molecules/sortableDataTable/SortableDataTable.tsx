import {
  DataTable,
  Table,
  TableHead,
  TableRow,
  TableHeader,
  TableBody,
  TableCell,
  DataTableHeader,
  DataTableSkeleton,
} from '@carbon/react';
import classnames from 'classnames';
import React from 'react';
import styles from './styles/SortableDataTable.module.scss';

interface SortableDataTableProps<T> {
  headers: DataTableHeader[];
  rows: T[];
  sortable?: { key: string; sortable: boolean }[];
  ariaLabel: string;
  loading?: boolean;
  errorStateMessage?: string | null;
  emptyStateMessage?: string;
  renderCell?: (row: T, cellId: string) => React.ReactNode;
  className?: string;
  onRowClick?: (row: T) => void;
}

export const SortableDataTable = <T extends { id: string }>({
  headers,
  rows,
  ariaLabel,
  sortable = headers.map((header) => ({ key: header.key, sortable: true })),
  loading = false,
  errorStateMessage: error = null,
  emptyStateMessage = 'No data available.',
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  renderCell = (row, cellId) => (row as any)[cellId],
  className = 'sortable-data-table',
  onRowClick,
}: SortableDataTableProps<T>) => {
  if (error) {
    return (
      <p
        data-testid="sortable-table-error"
        className={styles.sortableDataTableBodyEmpty}
      >
        {error}
      </p>
    );
  }

  if (loading) {
    return (
      <div data-testid="sortable-table-skeleton" className={className}>
        <DataTableSkeleton
          columnCount={headers.length}
          rowCount={5}
          showHeader={false}
          showToolbar={false}
          compact
          className={styles.sortableDataTableSkeleton}
        />
      </div>
    );
  }

  if (!rows || rows.length === 0) {
    return (
      <p
        data-testid="sortable-table-empty"
        className={styles.sortableDataTableBodyEmpty}
      >
        {emptyStateMessage}
      </p>
    );
  }

  const rowMap = new Map(rows.map((row) => [row.id, row]));

  return (
    <div
      className={classnames(className, styles.sortableDataTableBody)}
      data-testid="sortable-data-table"
    >
      <DataTable rows={rows} headers={headers} isSortable size="md">
        {({
          rows: tableRows,
          headers: tableHeaders,
          getHeaderProps,
          getRowProps,
          getTableProps,
        }) => (
          <Table {...getTableProps()} aria-label={ariaLabel} size="md">
            <TableHead>
              <TableRow>
                {tableHeaders.map((header) => {
                  const headerProps = getHeaderProps({
                    header,
                    isSortable:
                      sortable.find((s) => s.key === header.key)?.sortable ??
                      false,
                  });
                  return (
                    <TableHeader {...headerProps} key={header.key}>
                      {header.header}
                    </TableHeader>
                  );
                })}
              </TableRow>
            </TableHead>
            <TableBody>
              {tableRows.map((row) => {
                const originalRow = rowMap.get(row.id)!;
                return (
                  <TableRow
                    {...getRowProps({ row })}
                    key={row.id}
                    onClick={
                      onRowClick ? () => onRowClick(originalRow) : undefined
                    }
                    className={onRowClick ? styles.interactiveRow : ''}
                  >
                    {row.cells.map((cell) => (
                      <TableCell key={cell.id}>
                        {renderCell(originalRow, cell.info.header)}
                      </TableCell>
                    ))}
                  </TableRow>
                );
              })}
            </TableBody>
          </Table>
        )}
      </DataTable>
    </div>
  );
};
