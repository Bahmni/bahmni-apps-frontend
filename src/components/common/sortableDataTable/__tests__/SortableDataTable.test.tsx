import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { SortableDataTable } from '../SortableDataTable';
import { DataTableHeader } from '@carbon/react';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);

const mockHeaders: DataTableHeader[] = [
  { key: 'name', header: 'Medication' },
  { key: 'dosage', header: 'Dosage' },
  { key: 'instruction', header: 'Instruction' },
  { key: 'startDate', header: 'Start Date' },
  { key: 'orderDate', header: 'Order Date' },
  { key: 'orderedBy', header: 'Ordered By' },
  { key: 'quantity', header: 'Quantity' },
  { key: 'status', header: 'Status' },
];

const mockMedicationRows = [
  {
    id: '2c6a89df-64a5-4a99-a078-4c3fda0b2f15',
    name: 'Acetylsalicylic acid 150 mg',
    dosage: '1 Tablet | Twice a day | 2 days',
    dosageUnit: 'Tablet',
    instruction: 'Oral',
    startDate: '03/04/2025',
    orderDate: '03/04/2025',
    orderedBy: 'Super Man',
    quantity: '4 Tablet',
    status: 'stopped',
  },
  {
    id: 'fb4bc0c6-849c-4ec3-ad63-0e18b0ee3604',
    name: 'Paroxetine 12.5 mg',
    dosage: '1 Tablet | Once a day | 5 days',
    dosageUnit: 'Tablet',
    instruction: 'Nasal',
    startDate: '04/04/2025',
    orderDate: '04/04/2025',
    orderedBy: 'Dr Neha Anand',
    quantity: '5 Tablet',
    status: 'stopped',
  },
  {
    id: '1124a9bb-1891-44bb-9842-103239aa2ca3',
    name: 'Chlorpheniramine maleate 4 mg',
    dosage: '2 Tablet | Twice a day | 2 days',
    dosageUnit: 'Tablet',
    instruction: 'Injection',
    startDate: '08/04/2025',
    orderDate: '08/04/2025',
    orderedBy: 'Dr Franklin Richards',
    quantity: '8 Tablet',
    status: 'scheduled',
  },
  {
    id: '1d2d372e-0ec3-4544-accd-aa5743d1bf70',
    name: 'Paracetamol 650 mg',
    dosage: '2 Tablet | Immediately | 2 days',
    dosageUnit: 'Tablet',
    instruction: 'Eye Drops',
    startDate: '08/04/2025',
    orderDate: '08/04/2025',
    orderedBy: 'Dr Sebastian Adams',
    quantity: '4 Tablet',
    status: 'stopped',
  },
  {
    id: 'cb23207e-102d-480f-b4e7-6b7ee2084ae7',
    name: 'Hydrochlorothiazide 25 mg',
    dosage: '2 Tablet | Thrice a day | 2 days',
    dosageUnit: 'Tablet',
    instruction: 'Oral | Patient needs sufficient dose of clean water',
    startDate: '08/04/2025',
    orderDate: '08/04/2025',
    orderedBy: 'Dr Monica Smith',
    quantity: '12 Tablet',
    status: 'stopped',
  },
  {
    id: 'c5396e12-dbfe-4259-9e42-92a52a2761a8',
    name: 'Oxygen',
    dosage: '2 Unit | Twice a day | 2 days',
    dosageUnit: 'Unit',
    instruction: 'Nasogastric | Patient needs sufficient dose of clean oxygen',
    startDate: '24/06/2025',
    orderDate: '24/06/2025',
    orderedBy: 'Dr John Doe',
    quantity: '8 Unit',
    status: 'active',
  },
];

describe('SortableDataTable', () => {
  const renderCell = (
    row: (typeof mockMedicationRows)[number],
    cellId: string,
  ) => row[cellId as keyof typeof row];

  it('renders medication table with all rows', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={mockMedicationRows}
        ariaLabel="Medication Orders"
        renderCell={renderCell}
      />,
    );

    // Check some representative rows
    expect(screen.getByText('Paracetamol 650 mg')).toBeInTheDocument();
    expect(screen.getByText('Oxygen')).toBeInTheDocument();
    expect(screen.getByText('Super Man')).toBeInTheDocument();
    expect(screen.getByText('Oral')).toBeInTheDocument();
    expect(screen.getByText('8 Unit')).toBeInTheDocument();
  });

  it('renders sortable columns and sorts on header click (e.g., name)', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={mockMedicationRows}
        ariaLabel="Medication Orders"
        renderCell={renderCell}
      />,
    );

    const nameHeader = screen.getByText('Medication');
    fireEvent.click(nameHeader); // simulate sort click

    // All data should still be present (just verify it re-renders correctly)
    expect(screen.getByText('Oxygen')).toBeInTheDocument();
    expect(screen.getByText('Hydrochlorothiazide 25 mg')).toBeInTheDocument();
  });
  it('renders error state when errorStateMessage is provided', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={mockMedicationRows}
        errorStateMessage="Something failed"
        ariaLabel="Error Table"
      />,
    );

    const errorElement = screen.getByTestId('sortable-table-error');
    expect(errorElement).toBeInTheDocument();
    expect(errorElement.textContent).toBe('Something failed');
  });

  it('renders skeleton when loading is true', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={mockMedicationRows}
        loading
        ariaLabel="Loading Table"
      />,
    );

    const skeleton = screen.getByTestId('sortable-table-skeleton');
    expect(skeleton).toBeInTheDocument();
  });

  it('renders empty state when rows are empty', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={[]}
        ariaLabel="Empty Table"
      />,
    );

    const empty = screen.getByTestId('sortable-table-empty');
    expect(empty).toBeInTheDocument();
    expect(empty.textContent).toBe('No data available.');
  });

  it('renders empty state with custom emptyStateMessage', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={[]}
        emptyStateMessage="Nothing to show"
        ariaLabel="Custom Empty Table"
      />,
    );

    const empty = screen.getByTestId('sortable-table-empty');
    expect(empty.textContent).toBe('Nothing to show');
  });

  it('renders the table with default renderCell', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={mockMedicationRows}
        ariaLabel="Default RenderCell"
      />,
    );

    expect(screen.getByText('Paracetamol 650 mg')).toBeInTheDocument();
    expect(screen.getByText('active')).toBeInTheDocument();
  });

  it('handles undefined rows gracefully', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        rows={undefined as any}
        ariaLabel="Undefined Rows"
      />,
    );

    expect(screen.getByTestId('sortable-table-empty')).toBeInTheDocument();
  });

  it('handles null rows gracefully', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        rows={null as any}
        ariaLabel="Null Rows"
      />,
    );

    expect(screen.getByTestId('sortable-table-empty')).toBeInTheDocument();
  });

  it('applies custom className and CSS module class', () => {
    const { container } = render(
      <SortableDataTable
        headers={mockHeaders}
        rows={mockMedicationRows}
        className="my-custom-table"
        ariaLabel="Styled Table"
      />,
    );

    expect(container.firstChild).toHaveClass('my-custom-table');
  });

  it('respects partial sortable array', () => {
    render(
      <SortableDataTable
        headers={mockHeaders}
        rows={mockMedicationRows}
        sortable={[{ key: 'name', sortable: false }]} // status is not defined here
        ariaLabel="Partial Sortable"
      />,
    );

    const headers = screen.getAllByRole('columnheader');
    expect(headers[0].querySelector('button')).toBeNull(); // name not sortable
    // status not in `sortable` array, default = false
    expect(headers[1].querySelector('button')).toBeNull();
  });

  describe('Snapshots', () => {
    it('matches snapshot with full data', () => {
      const { container } = render(
        <SortableDataTable
          headers={mockHeaders}
          rows={mockMedicationRows}
          ariaLabel="Snapshot Test Full"
          renderCell={renderCell}
        />,
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot when loading', () => {
      const { container } = render(
        <SortableDataTable
          headers={mockHeaders}
          rows={mockMedicationRows}
          loading
          ariaLabel="Snapshot Loading"
        />,
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot when empty', () => {
      const { container } = render(
        <SortableDataTable
          headers={mockHeaders}
          rows={[]}
          emptyStateMessage="No medication history"
          ariaLabel="Snapshot Empty"
        />,
      );
      expect(container).toMatchSnapshot();
    });
  });
  describe('Accessibility', () => {
    it('has no accessibility violations', async () => {
      // Arrange
      const { container } = render(
        <SortableDataTable
          headers={mockHeaders}
          rows={mockMedicationRows}
          ariaLabel="Accessibility Test"
          renderCell={renderCell}
        />,
      );
      expect(await axe(container)).toHaveNoViolations();
    });
  });
});
