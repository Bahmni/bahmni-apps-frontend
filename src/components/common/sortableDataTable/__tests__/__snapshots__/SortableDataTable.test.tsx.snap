// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableDataTable Snapshots matches snapshot when empty 1`] = `
<div>
  <p
    data-testid="sortable-table-empty"
  >
    No medication history
  </p>
</div>
`;

exports[`SortableDataTable Snapshots matches snapshot when loading 1`] = `
<div>
  <div
    class="sortable-data-table"
    data-testid="sortable-table-skeleton"
  >
    <div
      class="cds--skeleton cds--data-table-container"
    >
      <table
        class="cds--skeleton cds--data-table cds--data-table--compact"
      >
        <thead>
          <tr>
            <th>
              <span />
            </th>
            <th>
              <span />
            </th>
            <th>
              <span />
            </th>
            <th>
              <span />
            </th>
            <th>
              <span />
            </th>
            <th>
              <span />
            </th>
            <th>
              <span />
            </th>
            <th>
              <span />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
          </tr>
          <tr>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
          </tr>
          <tr>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
          </tr>
          <tr>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
          </tr>
          <tr>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
            <td>
              <span />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`SortableDataTable Snapshots matches snapshot with full data 1`] = `
<div>
  <div
    class="sortable-data-table"
    data-testid="sortable-data-table"
  >
    <section
      class="cds--data-table-container"
    >
      <div
        class="cds--data-table-content"
      >
        <table
          aria-label="Snapshot Test Full"
          class="cds--data-table cds--data-table--lg cds--data-table--sort cds--data-table--visible-overflow-menu"
        >
          <thead>
            <tr>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1e»"
                >
                  Click to sort rows by Medication header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1e»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Medication
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1f»"
                >
                  Click to sort rows by Dosage header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1f»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Dosage
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1g»"
                >
                  Click to sort rows by Instruction header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1g»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Instruction
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1h»"
                >
                  Click to sort rows by Start Date header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1h»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Start Date
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1i»"
                >
                  Click to sort rows by Order Date header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1i»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Order Date
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1j»"
                >
                  Click to sort rows by Ordered By header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1j»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Ordered By
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1k»"
                >
                  Click to sort rows by Quantity header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1k»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Quantity
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="cds--table-sort__header"
                scope="col"
              >
                <div
                  class="cds--table-sort__description"
                  id="table-sort-«r1l»"
                >
                  Click to sort rows by Status header in ascending order
                </div>
                <button
                  aria-describedby="table-sort-«r1l»"
                  class="cds--table-sort"
                  type="button"
                >
                  <span
                    class="cds--table-sort__flex"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Status
                    </div>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="cds--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                    <div
                      class="cds--table-header-label--decorator-inner"
                    />
                  </span>
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            aria-live="polite"
          >
            <tr>
              <td>
                Acetylsalicylic acid 150 mg
              </td>
              <td>
                1 Tablet | Twice a day | 2 days
              </td>
              <td>
                Oral
              </td>
              <td>
                03/04/2025
              </td>
              <td>
                03/04/2025
              </td>
              <td>
                Super Man
              </td>
              <td>
                4 Tablet
              </td>
              <td>
                stopped
              </td>
            </tr>
            <tr>
              <td>
                Paroxetine 12.5 mg
              </td>
              <td>
                1 Tablet | Once a day | 5 days
              </td>
              <td>
                Nasal
              </td>
              <td>
                04/04/2025
              </td>
              <td>
                04/04/2025
              </td>
              <td>
                Dr Neha Anand
              </td>
              <td>
                5 Tablet
              </td>
              <td>
                stopped
              </td>
            </tr>
            <tr>
              <td>
                Chlorpheniramine maleate 4 mg
              </td>
              <td>
                2 Tablet | Twice a day | 2 days
              </td>
              <td>
                Injection
              </td>
              <td>
                08/04/2025
              </td>
              <td>
                08/04/2025
              </td>
              <td>
                Dr Franklin Richards
              </td>
              <td>
                8 Tablet
              </td>
              <td>
                scheduled
              </td>
            </tr>
            <tr>
              <td>
                Paracetamol 650 mg
              </td>
              <td>
                2 Tablet | Immediately | 2 days
              </td>
              <td>
                Eye Drops
              </td>
              <td>
                08/04/2025
              </td>
              <td>
                08/04/2025
              </td>
              <td>
                Dr Sebastian Adams
              </td>
              <td>
                4 Tablet
              </td>
              <td>
                stopped
              </td>
            </tr>
            <tr>
              <td>
                Hydrochlorothiazide 25 mg
              </td>
              <td>
                2 Tablet | Thrice a day | 2 days
              </td>
              <td>
                Oral | Patient needs sufficient dose of clean water
              </td>
              <td>
                08/04/2025
              </td>
              <td>
                08/04/2025
              </td>
              <td>
                Dr Monica Smith
              </td>
              <td>
                12 Tablet
              </td>
              <td>
                stopped
              </td>
            </tr>
            <tr>
              <td>
                Oxygen
              </td>
              <td>
                2 Unit | Twice a day | 2 days
              </td>
              <td>
                Nasogastric | Patient needs sufficient dose of clean oxygen
              </td>
              <td>
                24/06/2025
              </td>
              <td>
                24/06/2025
              </td>
              <td>
                Dr John Doe
              </td>
              <td>
                8 Unit
              </td>
              <td>
                active
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </section>
  </div>
</div>
`;
